power_grid_model_c/power_grid_model/include/power_grid_model/container_fwd.hpp

```diff
@@ -0,0 +1,32 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "common/common.hpp"
+
+#include <concepts>
+
+namespace power_grid_model {
+
+template <typename ContainerType, typename ComponentType>
+concept component_container_c = requires(ContainerType const& c, ID id) {
+    { c.template citer<ComponentType>().begin() } -> std::forward_iterator;
+    { c.template citer<ComponentType>().end() } -> std::forward_iterator;
+    { *(c.template citer<ComponentType>().begin()) } -> std::same_as<ComponentType const&>;
+    { *(c.template citer<ComponentType>().end()) } -> std::same_as<ComponentType const&>;
```

Comment:
```suggestion
```
This should not be present right since it makes no sense to dereference it. Is this correct?

Reply:
you are right in the sense that it is undefined behavior, though... Maybe you can replace it with something along the lines of


```suggestion
    { c.template citer<ComponentType>().end() } -> std::same_as<decltype(c.template citer<ComponentType>().begin())>;
```